### PR TITLE
MRG: Vendor simplified PCA

### LINKED ISSUE #6062: Do we need scikit-learn for Infomax ICA?
Currently, we [require `scikit-learn` for all ICA algorithms](https://github.com/mne-tools/mne-python/blob/cffcf4b1aeaf04655f6917718029c59ff412c917/mne/preprocessing/ica.py#L316). However, I don't think that Infomax requires `scikit-learn`, only FastICA uses the implementation in this package. Not sure how we handle Picard, but I think it requires `scikit-learn` either.

### PR DESCRIPTION
**Pull request by larsoner**
Closes #6062.

### CODE REVIEW

File: mne/fixes.py [hunk 1]
```diff
@@ -1015,6 +1016,95 @@ def _logdet(A):
...
+    for i in range(rank):
+        pu += (gammaln((n_features - i) / 2.) -
+               log(np.pi) * (n_features - i) / 2.)
+
+    pl = np.sum(np.log(spectrum[:rank]))
+    pl = -pl * n_samples / 2.
+
+    if rank == n_features:
+        pv = 0
+        v = 1
```
Comment (jona-sassenhagen) [inline mne/fixes.py]: I guess you don't want to go `pv, v = 0, 1` for symmetry?

But in that case, make this

```
v = 1
pv = 0
```

as below?

File: mne/fixes.py [hunk 2]
```diff
@@ -1015,6 +1016,95 @@ def _logdet(A):
...
+        v = np.sum(spectrum[rank:]) / (n_features - rank)
+        pv = -np.log(v) * n_samples * (n_features - rank) / 2.
+
+    m = n_features * rank - rank * (rank + 1.) / 2.
+    pp = log(2. * np.pi) * (m + rank + 1.) / 2.
+
+    pa = 0.
+    spectrum_ = spectrum.copy()
+    spectrum_[rank:n_features] = v
+    for i in range(rank):
```
Comment (jona-sassenhagen) [inline mne/fixes.py]: We don't iterate over single-char variables usually ...

File: mne/fixes.py [hunk 3]
```diff
@@ -1015,6 +1016,95 @@ def _logdet(A):
...
+        max_abs_cols = np.argmax(np.abs(u), axis=0)
+        signs = np.sign(u[max_abs_cols, np.arange(u.shape[1])])
+        u *= signs
+        v *= signs[:, np.newaxis]
+    else:
+        # rows of v, columns of u
+        max_abs_rows = np.argmax(np.abs(v), axis=1)
+        signs = np.sign(v[np.arange(v.shape[0]), max_abs_rows])
+        u *= signs
+        v *= signs[:, np.newaxis]
```
Comment (jona-sassenhagen) [inline mne/fixes.py]: Ok, this code is just copied from somewhere right?

1072, 1073 and 1078, 1079 should go outside of the loop.

File: mne/utils/numerics.py [hunk 4]
```diff
@@ -738,3 +741,95 @@ def object_diff(a, b, pre=''):
...
     return out
+
+
+class _PCA(object):
+    """Principal component analysis (PCA)."""
+
+    # Adapted from sklearn and stripped down to just use linalg.svd
+    # and make it easier to later provide a "center" option if we want
+
+    def __init__(self, n_components=None, copy=True, whiten=False):
```
Comment (jona-sassenhagen) [inline mne/utils/numerics.py]: Do we need the copy parameter? We always copy, right?

File: mne/fixes.py (hunk 3 above)
Comment (larsoner) [reply mne/fixes.py]: Things in fixes should stay as close to verbatim copies as reasonable

Comment (jona-sassenhagen) [reply mne/fixes.py]: Mh, ok.

**Comment by larsoner** [comment]
Pushed a commit to  remove `copy=True`, since it's the `sklearn` default anyway I agree it's unnecsesary. Leaving `pca=False`, by contrast, makes sense since the default in `sklearn` is `False` (even though we currently always in ICA use `whiten=True`). We test both modes in the `numerics` equivalence tests at least.

I also moved `mne/utils/testing.py` to `mne/utils/_testing.py` because in `master` tab-completion for the `mne/utils/tests` subdirectory when trying to get to some specific `test_*.py` file gets stuck at `mne/utils/test` due to overlap.

**Comment by agramfort** [comment]
is it all copy pasted from sklearn code?

**Comment by larsoner** [comment]
Copy-pasted and trimmed down, yes

**Comment by agramfort** [comment]
+1 for MRG then

we can add the center option later.

>

**Comment by larsoner** [comment]
@cbrnr feel free to merge if you're happy

**Comment by cbrnr** [comment]
Very nice! Thanks @larsoner!

**Pull request merged**